### PR TITLE
Fixing uap/uapaot test issues in System.IO.Compression.ZipFile

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes two System.IO.Compression.ZipFile tempdir issues. Using temppath instead of directly reading from execution location.

https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fuwp~2F/build/20170513.02/workItem/System.IO.Compression.ZipFile.Tests

### CODE REVIEW

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs [hunk 1]
```diff
@@ -46,55 +46,85 @@ public void InvalidConstructors()
         [Fact]
         public void InvalidFiles()
         {
-            ConstructorThrows<InvalidDataException>(() => ZipFile.OpenRead(bad("EOCDmissing.zip")));
-            ConstructorThrows<InvalidDataException>(() => ZipFile.Open(bad("EOCDmissing.zip"), ZipArchiveMode.Update));
-            ConstructorThrows<InvalidDataException>(() => ZipFile.OpenRead(bad("CDoffsetOutOfBounds.zip")));
-            ConstructorThrows<InvalidDataException>(() => ZipFile.Open(bad("CDoffsetOutOfBounds.zip"), ZipArchiveMode.Update));
+            using (TempFile testArchive = CreateTempCopyFile(bad("EOCDmissing.zip"), GetTestFilePath()))
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: Why did you change the ones the were only reading? That shouldn't be necessary.

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs [hunk 2]
```diff
@@ -46,55 +46,85 @@ public void InvalidConstructors()
...
+            {
+                ConstructorThrows<InvalidDataException>(() => ZipFile.OpenRead(testArchive.Path));
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: I don't know why this ConstructorThrows helper is needed... if the ctor throws, there's nothing to dispose, and having the dispose in the Throws as well opens the possibility that we're actually catching the exception from Dispose rather than the ctor.

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs (hunk 1 above)
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]:
```
  ERROR:    System.IO.Compression.Tests.ZipFileTest_Invalid.InvalidFiles [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.IO.InvalidDataException)
        Actual:   typeof(System.UnauthorizedAccessException): Access to the path 'C:\Users\vikto\Documents\src\corefx\b
  in\AnyOS.AnyCPU.Debug\System.IO.Compression.ZipFile.Tests\netstandard\ZipTestData\badzipfiles\EOCDmissing.zip' is den
  ied.
```

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs (hunk 2 above)
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: You are right. There is already a catch in the ZipFile.Open method which disposes the FileStream if something fails. https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs#L205

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs (hunk 1 above)
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: I see. I didn't notice that not all of them have `ZipArchiveMode.Update` as a param. Thanks!

File: src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs (hunk 2 above)
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: You don't need it where Zip.Open/ZipFile.OpenRead is used, as it already disposes the FileStream if something fails: https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs#L205

But e.g. in this line https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs#L67 where ZipArchive constructor is directly called I think you need it as the constructor calls the Init function which opens multiple streams which need to be disposed in case of an error. Right?

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: I updated the changes.

Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: > you need it as the constructor of ZipArchive opens multiple streams which need to be disposed in case of an error.

If it throws, your don't get anything back to dispose. If it doesn't throw, the test fails. What an I missing?

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: I reference to this post: http://stackoverflow.com/a/30990784/2415001

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: But yeah, you are perfectly right that the using statement isn't doing anything and should be removed as there is no object created which could be disposed (from the outside). I wanted to point out that I think we are not releasing all resources inside ZipArchive which are getting creating during the construction.

Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: > Correct, which explains why the ConstructorThrows helper isn't necessary, 

Right

> We could argue that ZipArchive should clean up itself if it fails...

I've not looked at the code... if it's opening streams and then leaving them open in the case of failure, that's a bug.

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: Alright. I took another look, and it seems, all streams which are opened inside ZipArchive are being closed correctly. The only thing which makes we wonder is, what this Stream is created for: https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L426

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: > The only thing which makes we wonder is, what this Stream is created for

I now know why it is created. If the stream is not seekable the stream param get set to a new Stream which changes the pointer address. So only if this new Stream get created it should be disposed. It's just a check if a new Stream got created.

Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFileInvalidFileTests.cs]: @stephentoub my changes should be fine now.
